File: src/pratice/binarysearch/k번째수1300.java
package pratice.binarysearch;

import java.util.Scanner;

public class k번째수1300 {
    public static int N;
    public static int M;
    public static void main(String[] args) {
        input();
        pro();
    }

    private static void pro() {
        long L = 1;
        long R = (long)N * N;
        long res = 0;
        while(L <= R){
            long mid = (L + R) / 2;
            if(isAble(mid)){
                res = mid;
                R = mid - 1;
            }else{
                L = mid + 1;
            }
        }
        System.out.println(res);
    }

    private static boolean isAble(long candidate) {
        long sum = 0;
        for (int i = 1; i <= N ; i++) {
            sum += Math.min(N, candidate / i);
        }
        return sum >= M;
    }

    private static void input() {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        M = sc.nextInt();
        // 가운데 배열 삽입
    }
}

File: src/pratice/binarysearch/K번째수1300.kt
package pratice.binarysearch

import kotlin.math.min

fun main() {
    val (N, K) = input()
    pro(N, K)
}

private fun pro(N: Int, K: Int) {
    var L = 1L
    var R: Long = (N.toLong() * N)
    var res = 0L
    while (L <= R) {
        val mid: Long = (L + R) / 2
        if (determination(N, K, mid)) {
            res = mid
            R = mid - 1
        } else {
            L = mid + 1
        }
    }
    println(res)
}

private fun input(): List<Int> {
    val N = readln().toInt()
    val K = readln().toInt()
    return listOf(N, K)
}

private fun determination(N: Int, K: Int, mid: Long): Boolean {
    var cnt = 0L
    for (i in 1..N) {
        cnt += min(N.toLong(), mid / i)
    }
    return cnt >= K
}
